File: AI/memory/dual_memory.py
"""
Dual Memory System
双层记忆系统：短期记忆（滑动窗口）+ 长期记忆（向量数据库）
"""

from typing import List, Dict, Optional, Any
from dataclasses import dataclass, field
from langchain_core.messages import BaseMessage, HumanMessage, AIMessage, SystemMessage
import uuid
import json


@dataclass
class MemoryItem:
    """记忆项"""
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    content: str = ""
    importance: float = 0.5  # 重要性评分 0-1
    timestamp: float = 0.0
    metadata: Dict = field(default_factory=dict)
    
    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "content": self.content,
            "importance": self.importance,
            "timestamp": self.timestamp,
            "metadata": self.metadata,
        }


class ShortTermMemory:
    """
    短期记忆 - 基于滑动窗口
    
    保留最近 N 轮对话，超出时自动压缩
    """
    
    def __init__(self, window_size: int = 5):
        """
        初始化短期记忆
        
        Args:
            window_size: 滑动窗口大小（保留的对话轮数）
        """
        self.window_size = window_size
        self.messages: List[BaseMessage] = []
        self.summary: str = ""  # 历史对话摘要
        self.conversation_rounds: int = 0
        
    def add_message(self, message: BaseMessage) -> None:
        """添加消息"""
        self.messages.append(message)
        
        if isinstance(message, HumanMessage):
            self.conversation_rounds += 1
            
    def get_messages(self) -> List[BaseMessage]:
        """获取当前窗口内的消息"""
        # 如果消息数量超过窗口大小，需要压缩
        if len(self.messages) > self.window_size * 2:  # 每轮 2 条消息
            # 保留最近的窗口大小
            overflow = self.messages[:len(self.messages) - self.window_size * 2]
            self._compress(overflow)
            
        return self.messages
    
    def _compress(self, messages_to_compress: List[BaseMessage]) -> None:
        """
        压缩消息为摘要
        
        Args:
            messages_to_compress: 需要压缩的消息列表
        """
        # TODO: 实际实现中应该调用 LLM 进行压缩
        # 这里先简单拼接
        compressed_text = "\n".join([
            f"{type(m).__name__}: {m.content}" 
            for m in messages_to_compress
        ])
        
        if self.summary:
            self.summary += f"\n[旧对话]\n{compressed_text}"
        else:
            self.summary = f"[历史对话摘要]\n{compressed_text}"
        
        # 移除已压缩的消息
        self.messages = self.messages[len(messages_to_compress):]
    
    def get_context(self) -> str:
        """获取完整的上下文（摘要 + 原始消息）"""
        context_parts = []
        
        if self.summary:
            context_parts.append(self.summary)
        
        for msg in self.messages:
            if isinstance(msg, HumanMessage):
                context_parts.append(f"用户：{msg.content}")
            elif isinstance(msg, AIMessage):
                context_parts.append(f"助手：{msg.content}")
            elif isinstance(msg, SystemMessage):
                context_parts.append(f"系统：{msg.content}")
        
        return "\n".join(context_parts)
    
    def clear(self) -> None:
        """清空短期记忆"""
        self.messages.clear()
        self.summary = ""
        self.conversation_rounds = 0
    
    def get_state(self) -> Dict:
        """获取状态"""
        return {
            "window_size": self.window_size,
            "message_count": len(self.messages),
            "conversation_rounds": self.conversation_rounds,
            "has_summary": bool(self.summary),
        }


class LongTermMemory:
    """
    长期记忆 - 基于向量数据库
    
    使用 ChromaDB/FAISS 存储重要对话和信息，支持语义检索
    """
    
    def __init__(
        self,
        storage_path: str = "./memory_storage",
        collection_name: str = "long_term_memory",
        use_mock: bool = True,  # 暂时使用模拟实现
    ):
        """
        初始化长期记忆
        
        Args:
            storage_path: 存储路径
            collection_name: 集合名称
            use_mock: 是否使用模拟实现（后续替换为真实的 ChromaDB）
        """
        self.storage_path = storage_path
        self.collection_name = collection_name
        self.use_mock = use_mock
        
        # 模拟存储
        self._mock_store: List[MemoryItem] = []
        
        # TODO: 真实实现时初始化 ChromaDB
        # self.client = chromadb.PersistentClient(path=storage_path)
        # self.collection = self.client.get_or_create_collection(name=collection_name)
    
    def add_memory(
        self,
        content: str,
        importance: float = 0.5,
        metadata: Optional[Dict] = None,
    ) -> str:
        """
        添加长期记忆
        
        Args:
            content: 记忆内容
            importance: 重要性评分 (0-1)
            metadata: 元数据
            
        Returns:
            记忆 ID
        """
        memory = MemoryItem(
            content=content,
            importance=importance,
            metadata=metadata or {},
        )
        
        if self.use_mock:
            self._mock_store.append(memory)
        else:
            # TODO: 真实实现：存入 ChromaDB
            pass
        
        return memory.id
    
    def search(
        self,
        query: str,
        top_k: int = 3,
        min_importance: float = 0.0,
    ) -> List[MemoryItem]:
        """
        检索记忆
        
        Args:
            query: 查询文本
            top_k: 返回数量
            min_importance: 最小重要性阈值
            
        Returns:
            相关记忆列表
        """
        if self.use_mock:
            # 模拟实现：简单关键词匹配
            results = []
            for memory in self._mock_store:
                if memory.importance >= min_importance:
                    if query.lower() in memory.content.lower():
                        results.append(memory)
            
            # 按重要性排序
            results.sort(key=lambda x: x.importance, reverse=True)
            return results[:top_k]
        else:
            # TODO: 真实实现：使用 ChromaDB 语义检索
            return []
    
    def get_all_memories(self) -> List[MemoryItem]:
        """获取所有记忆"""
        if self.use_mock:
            return self._mock_store.copy()
        return []
    
    def count(self) -> int:
        """获取记忆数量"""
        if self.use_mock:
            return len(self._mock_store)
        return 0
    
    def clear(self) -> None:
        """清空长期记忆"""
        if self.use_mock:
            self._mock_store.clear()


class DualMemorySystem:
    """
    双层记忆系统
    
    整合短期和长期记忆，提供统一的接口
    """
    
    def __init__(
        self,
        agent_id: str,
        short_term_window: int = 5,
        long_term_storage_path: str = "./memory_storage",
    ):
        """
        初始化双层记忆系统
        
        Args:
            agent_id: 智能体 ID
            short_term_window: 短期记忆窗口大小
            long_term_storage_path: 长期记忆存储路径
        """
        self.agent_id = agent_id
        self.short_term = ShortTermMemory(window_size=short_term_window)
        self.long_term = LongTermMemory(storage_path=long_term_storage_path)
        
        # 重要性阈值（高于此值才写入长期记忆）
        self.importance_threshold = 0.6
    
    def add_message(
        self,
        message: BaseMessage,
        evaluate_importance: bool = True,
    ) -> None:
        """
        添加消息到记忆系统
        
        Args:
            message: 消息对象
            evaluate_importance: 是否评估重要性
        """
        # 添加到短期记忆
        self.short_term.add_message(message)
        
        # 评估是否重要到需要存入长期记忆
        if evaluate_importance and isinstance(message, (HumanMessage, AIMessage)):
            importance = self._evaluate_importance(message.content)
            
            if importance >= self.importance_threshold:
                self.long_term.add_memory(
                    content=message.content,
                    importance=importance,
                    metadata={
                        "type": type(message).__name__,
                        "agent_id": self.agent_id,
                    }
                )
    
    def _evaluate_importance(self, content: str) -> float:
        """
        评估内容的重要性
        
        Args:
            content: 内容文本
            
        Returns:
            重要性评分 (0-1)
        """
        # TODO: 实际应该用 LLM 评估
        # 简单规则：包含特定关键词的更重要
        important_keywords = ["重要", "记住", "关键", "决定", "承诺", "秘密"]
        
        score = 0.3  # 基础分
        for keyword in important_keywords:
            if keyword in content:
                score += 0.15
        
        return min(score, 1.0)
    
    def get_context(
        self,
        query: Optional[str] = None,
        include_long_term: bool = True,
    ) -> str:
        """
        获取完整的记忆上下文
        
        Args:
            query: 查询文本（用于检索长期记忆）
            include_long_term: 是否包含长期记忆
            
        Returns:
            格式化的上下文文本
        """
        context_parts = []
        
        # 1. 短期记忆（总是包含）
        short_term_context = self.short_term.get_context()
        if short_term_context:
            context_parts.append(short_term_context)
        
        # 2. 长期记忆（根据查询检索）
        if include_long_term and query:
            relevant_memories = self.long_term.search(query, top_k=3)
            
            if relevant_memories:
                memories_text = "\n".join([
                    f"- {mem.content} (重要性：{mem.importance:.2f})"
                    for mem in relevant_memories
                ])
                context_parts.append(f"\n[相关长期记忆]\n{memories_text}")
        
        return "\n\n".join(context_parts)
    
    def save_important_event(
        self,
        event_description: str,
        importance: float = 0.8,
    ) -> str:
        """
        保存重要事件到长期记忆
        
        Args:
            event_description: 事件描述
            importance: 重要性评分
            
        Returns:
            记忆 ID
        """
        return self.long_term.add_memory(
            content=event_description,
            importance=importance,
            metadata={"type": "event"}
        )
    
    def get_state(self) -> Dict:
        """获取记忆系统状态"""
        return {
            "agent_id": self.agent_id,
            "short_term": self.short_term.get_state(),
            "long_term_count": self.long_term.count(),
            "importance_threshold": self.importance_threshold,
        }
    
    def clear(self) -> None:
        """清空所有记忆"""
        self.short_term.clear()
        self.long_term.clear()


def create_memory_system(
    agent_id: str,
    window_size: int = 5,
    **kwargs
) -> DualMemorySystem:
    """
    工厂函数：创建使用内存 Mock 的记忆系统（开发/测试用）

    Args:
        agent_id: 智能体 ID
        window_size: 短期记忆窗口大小
        **kwargs: 其他参数

    Returns:
        DualMemorySystem 实例
    """
    return DualMemorySystem(
        agent_id=agent_id,
        short_term_window=window_size,
        **kwargs
    )


def create_pg_memory_system(
    agent_id: str,
    window_size: int = 5,
    collection_name: str = "default",
) -> DualMemorySystem:
    """
    工厂函数：创建使用 PostgreSQL 长期记忆的记忆系统（生产用）

    长期记忆使用 PGLongTermMemory 替代内存 Mock。
    需要在使用前调用 await system.long_term.initialize() 初始化连接池。

    Args:
        agent_id: 智能体 ID
        window_size: 短期记忆窗口大小
        collection_name: PG 记忆集合名称

    Returns:
        DualMemorySystem 实例（长期记忆为 PGLongTermMemory）
    """
    from .pg_long_term_memory import PGLongTermMemory

    system = DualMemorySystem(
        agent_id=agent_id,
        short_term_window=window_size,
    )
    # 替换长期记忆后端为 PostgreSQL
    system.long_term = PGLongTermMemory(
        agent_id=agent_id,
        collection_name=collection_name,
    )
    return system

File: src/AI/memory/dual_memory.ts
// 导入必要的类和接口
import { BaseMessage, ToolResultMessage } from '../agents/base_agent';
import { SessionMemory } from './session_memory';
import { PGLongTermMemory } from './pg_long_term_memory';
import { ChromaLongTermMemory } from './chroma_long_term_memory';
import { TokenUtils } from './token_utils';

/**
 * 记忆项接口，定义了记忆的基本属性和方法
 */
export interface MemoryItem {
  id: string;                    // 记忆ID
  content: string;               // 记忆内容
  importance: number;            // 记忆重要性（0-1）
  timestamp: number;             // 记忆时间戳
  metadata: Record<string, any>;  // 记忆元数据
  toDict(): Record<string, any>;  // 转换为字典的方法
}

/**
 * 长期记忆接口，定义了长期记忆的基本操作
 */
export interface LongTermMemoryInterface {
  addMemory(content: string, importance?: number, metadata?: Record<string, any>): Promise<string>; // 添加记忆
  search(query: string, topK?: number, minImportance?: number): Promise<MemoryItem[]>; // 搜索记忆
  getAllMemories(): Promise<MemoryItem[]>; // 获取所有记忆
  count(): Promise<number>; // 获取记忆数量
  clear(): Promise<void>; // 清空记忆
}

/**
 * 记忆项实现类
 */
export class MemoryItemImpl implements MemoryItem {
  /**
   * 构造函数
   * @param id 记忆ID（默认自动生成）
   * @param content 记忆内容（默认空字符串）
   * @param importance 记忆重要性（默认0.5）
   * @param timestamp 记忆时间戳（默认当前时间）
   * @param metadata 记忆元数据（默认空对象）
   */
  constructor(
    public id: string = Math.random().toString(16).substring(2, 10),
    public content: string = "",
    public importance: number = 0.5,
    public timestamp: number = Date.now() / 1000,
    public metadata: Record<string, any> = {}
  ) {}

  /**
   * 转换为字典
   * @returns 记忆项的字典表示
   */
  toDict(): Record<string, any> {
    return {
      id: this.id,
      content: this.content,
      importance: this.importance,
      timestamp: this.timestamp,
      metadata: this.metadata
    };
  }
}

/**
 * 短期记忆类，存储最近的对话消息
 */
export class ShortTermMemory {
  private windowSize: number; // 窗口大小
  private messages: BaseMessage[]; // 消息列表
  private summary: string; // 对话摘要
  private conversationRounds: number; // 对话轮数
  private toolResultsLRU: Map<string, { message: BaseMessage, timestamp: number }>; // 工具结果LRU缓存
  private maxToolResults: number = 5; // 最大工具结果数量

  /**
   * 构造函数
   * @param windowSize 窗口大小（默认5）
   */
  constructor(windowSize: number = 5) {
    this.windowSize = windowSize;
    this.messages = [];
    this.summary = "";
    this.conversationRounds = 0;
    this.toolResultsLRU = new Map<string, { message: BaseMessage, timestamp: number }>();
  }

  /**
   * 添加消息
   * @param message 消息对象
   */
  addMessage(message: BaseMessage): void {
    this.messages.push(message);
    if (message.type === 'human') {
      this.conversationRounds++;
    } else if (message.type === 'tool_result') {
      this.updateToolResultsLRU(message);
      this.performLocalCleanup();
    }
  }

  /**
   * 更新工具结果LRU缓存
   * @param message 工具结果消息
   */
  private updateToolResultsLRU(message: BaseMessage): void {
    const toolName = message.metadata?.tool_name || 'unknown';
    const timestamp = Date.now();
    const key = `${toolName}_${timestamp}`;

    // 检查键是否已存在（时间戳相同的边缘情况）
    if (this.toolResultsLRU.has(key)) {
      this.toolResultsLRU.delete(key);
    }

    // 添加到LRU缓存
    this.toolResultsLRU.set(key, { message, timestamp });

    // 如果超过最大大小，删除最旧的条目
    if (this.toolResultsLRU.size > this.maxToolResults) {
      // 查找并删除最旧的条目
      let oldestKey: string | null = null;
      let oldestTimestamp = Infinity;

      const entries = Array.from(this.toolResultsLRU.entries());
      for (const [k, v] of entries) {
        if (v.timestamp < oldestTimestamp) {
          oldestTimestamp = v.timestamp;
          oldestKey = k;
        }
      }

      if (oldestKey) {
        this.toolResultsLRU.delete(oldestKey);
      }
    }
  }

  /**
   * 执行本地清理
   */
  private performLocalCleanup(): void {
    // 从Map中获取值，避免Set引用比较问题
    const toolResultsToKeep = Array.from(this.toolResultsLRU.values()).map(entry => entry.message);

    // 过滤消息，将旧的工具结果替换为占位符
    this.messages = this.messages.map(msg => {
      if (msg.type === 'tool_result') {
        const shouldKeep = toolResultsToKeep.some(kept => kept === msg);
        if (!shouldKeep) {
          // 替换为占位符
          return new ToolResultMessage('[Old tool result content cleared]', msg.metadata);
        }
      }
      return msg;
    });
  }

  /**
   * 获取消息列表
   * @returns 消息列表
   */
  getMessages(): BaseMessage[] {
    const totalTokens = this.calculateTokenCount(this.messages);
    const maxTokens = this.windowSize * 1000; // 估计：每个窗口单位1000个token

    if (totalTokens > maxTokens * 1.5) {
      const overflow = this.messages.slice(0, this.messages.length - this.windowSize);
      this.compress(overflow);
    }
    return this.messages;
  }

  /**
   * 计算token数量
   * @param messages 消息列表
   * @returns token数量
   */
  private calculateTokenCount(messages: BaseMessage[]): number {
    return TokenUtils.calculateMessagesTokenCount(messages);
  }

  /**
   * 压缩消息
   * @param messagesToCompress 要压缩的消息
   */
  private compress(messagesToCompress: BaseMessage[]): void {
    const compressedText = messagesToCompress.map(msg => {
      return `${msg.type}: ${msg.content}`;
    }).join('\n');

    if (this.summary) {
      this.summary += `\n[旧对话]\n${compressedText}`;
    } else {
      this.summary = `[历史对话摘要]\n${compressedText}`;
    }

    this.messages = this.messages.slice(messagesToCompress.length);
  }

  /**
   * 获取上下文
   * @returns 上下文字符串
   */
  getContext(): string {
    const contextParts: string[] = [];

    if (this.summary) {
      contextParts.push(this.summary);
    }

    for (const msg of this.messages) {
      if (msg.type === 'human') {
        contextParts.push(`用户：${msg.content}`);
      } else if (msg.type === 'ai') {
        contextParts.push(`助手：${msg.content}`);
      } else if (msg.type === 'system') {
        contextParts.push(`系统：${msg.content}`);
      } else if (msg.type === 'tool') {
        contextParts.push(`工具：${msg.content}`);
      } else if (msg.type === 'tool_result') {
        contextParts.push(`工具结果：${msg.content}`);
      }
    }

    return contextParts.join('\n');
  }

  /**
   * 清空短期记忆
   */
  clear(): void {
    this.messages = [];
    this.summary = "";
    this.conversationRounds = 0;
    this.toolResultsLRU.clear();
  }

  /**
   * 获取短期记忆状态
   * @returns 状态对象
   */
  getState(): Record<string, any> {
    return {
      window_size: this.windowSize,
      message_count: this.messages.length,
      conversation_rounds: this.conversationRounds,
      has_summary: Boolean(this.summary),
      tool_results_count: this.toolResultsLRU.size
    };
  }
}

/**
 * 长期记忆类，实现了LongTermMemoryInterface接口
 */
export class LongTermMemory implements LongTermMemoryInterface {
  private storagePath: string; // 存储路径
  private collectionName: string; // 集合名称
  private useMock: boolean; // 是否使用模拟存储
  private mockStore: MemoryItem[]; // 模拟存储

  /**
   * 构造函数
   * @param storagePath 存储路径（默认"./memory_storage"）
   * @param collectionName 集合名称（默认"long_term_memory"）
   * @param useMock 是否使用模拟存储（默认true）
   */
  constructor(
    storagePath: string = "./memory_storage",
    collectionName: string = "long_term_memory",
    useMock: boolean = true
  ) {
    this.storagePath = storagePath;
    this.collectionName = collectionName;
    this.useMock = useMock;
    this.mockStore = [];
  }

  /**
   * 添加记忆
   * @param content 记忆内容
   * @param importance 记忆重要性（默认0.5）
   * @param metadata 记忆元数据（可选）
   * @returns 记忆ID
   */
  async addMemory(
    content: string,
    importance: number = 0.5,
    metadata?: Record<string, any>
  ): Promise<string> {
    const memory = new MemoryItemImpl(
      undefined,
      content,
      importance,
      undefined,
      metadata || {}
    );

    if (this.useMock) {
      this.mockStore.push(memory);
    } else {
      // TODO: 真实实现：存入向量数据库
    }

    return memory.id;
  }

  /**
   * 搜索记忆
   * @param query 搜索查询
   * @param topK 返回结果数量（默认3）
   * @param minImportance 最小重要性（默认0.0）
   * @returns 相关记忆数组
   */
  async search(
    query: string,
    topK: number = 3,
    minImportance: number = 0.0
  ): Promise<MemoryItem[]> {
    if (this.useMock) {
      const results: MemoryItem[] = [];
      for (const memory of this.mockStore) {
        if (memory.importance >= minImportance) {
          if (memory.content.toLowerCase().includes(query.toLowerCase())) {
            results.push(memory);
          }
        }
      }

      results.sort((a, b) => b.importance - a.importance);
      return results.slice(0, topK);
    } else {
      // TODO: 真实实现：使用向量数据库语义检索
      return [];
    }
  }

  /**
   * 获取所有记忆
   * @returns 所有记忆数组
   */
  async getAllMemories(): Promise<MemoryItem[]> {
    if (this.useMock) {
      return [...this.mockStore];
    }
    return [];
  }

  /**
   * 获取记忆数量
   * @returns 记忆数量
   */
  async count(): Promise<number> {
    if (this.useMock) {
      return this.mockStore.length;
    }
    return 0;
  }

  /**
   * 清空记忆
   */
  async clear(): Promise<void> {
    if (this.useMock) {
      this.mockStore = [];
    }
  }
}

/**
 * 双记忆系统类，结合短期记忆和长期记忆
 */
export class DualMemorySystem {
  private agentId: string; // 智能体ID
  private shortTerm: ShortTermMemory; // 短期记忆
  private longTerm: LongTermMemoryInterface; // 长期记忆
  private sessionMemory: SessionMemory; // 会话记忆
  private importanceThreshold: number; // 重要性阈值

  /**
   * 构造函数
   * @param agentId 智能体ID
   * @param shortTermWindow 短期记忆窗口大小（默认5）
   * @param longTermStoragePath 长期记忆存储路径（默认"./chroma"）
   */
  constructor(
    agentId: string,
    shortTermWindow: number = 5,
    longTermStoragePath: string = "./chroma"
  ) {
    this.agentId = agentId;
    this.shortTerm = new ShortTermMemory(shortTermWindow);
    // 使用ChromaDB作为长期记忆存储
    this.longTerm = new ChromaLongTermMemory(agentId, "long_term_memory", longTermStoragePath);
    this.sessionMemory = new SessionMemory(agentId, longTermStoragePath);
    this.importanceThreshold = 0.6;
  }

  /**
   * 设置长期记忆
   * @param longTerm 长期记忆实例
   */
  setLongTermMemory(longTerm: LongTermMemoryInterface): void {
    this.longTerm = longTerm;
  }

  /**
   * 添加消息
   * @param message 消息对象
   * @param evaluateImportance 是否评估重要性（默认true）
   */
  async addMessage(
    message: BaseMessage,
    evaluateImportance: boolean = true
  ): Promise<void> {
    this.shortTerm.addMessage(message);

    if (evaluateImportance && (message.type === 'human' || message.type === 'ai')) {
      const importance = this.evaluateImportance(message.content);

      if (importance >= this.importanceThreshold) {
        await this.longTerm.addMemory(
          message.content,
          importance,
          {
            type: message.type,
            agent_id: this.agentId
          }
        );
      }
    }

    // 检查是否需要提取会话记忆
    const messages = this.shortTerm.getMessages();
    const tokenCount = this.calculateTokenCount(messages);
    if (this.sessionMemory.shouldExtractMemory(messages, tokenCount)) {
      // 等待提取记忆以确保数据一致性
      await this.sessionMemory.extractMemory(messages);
    }
  }

  /**
   * 评估重要性
   * @param content 内容
   * @returns 重要性分数（0-1）
   */
  private evaluateImportance(content: string): number {
    // 支持中英文关键词
    const importantKeywords = [
      // 中文
      "重要", "记住", "关键", "决定", "承诺", "秘密", "必须", "一定要",
      // 英文
      "important", "remember", "key", "critical", "decision", "promise", "secret", "must"
    ];

    const contentLower = content.toLowerCase();
    let score = 0.3; // 基础分数

    for (const keyword of importantKeywords) {
      if (contentLower.includes(keyword.toLowerCase())) {
        score += 0.15;
      }
    }

    // 额外评分因素
    const hasNumbers = /\d+/.test(content);
    const hasUrls = /https?:\/\/|www\./.test(content);
    const hasCode = /```|function |class |const |let |var /.test(content);

    if (hasNumbers) score += 0.05;
    if (hasUrls) score += 0.1;
    if (hasCode) score += 0.15;

    // 内容长度因素（较长的消息往往更重要）
    if (content.length > 200) score += 0.05;
    if (content.length > 500) score += 0.1;

    return Math.min(score, 1.0);
  }

  /**
   * 计算token数量
   * @param messages 消息列表
   * @returns token数量
   */
  private calculateTokenCount(messages: BaseMessage[]): number {
    return TokenUtils.calculateMessagesTokenCount(messages);
  }

  /**
   * 获取上下文
   * @param query 查询（可选）
   * @param includeLongTerm 是否包含长期记忆（默认true）
   * @returns 上下文字符串
   */
  async getContext(
    query?: string,
    includeLongTerm: boolean = true
  ): Promise<string> {
    const contextParts: string[] = [];

    const shortTermContext = this.shortTerm.getContext();
    if (shortTermContext) {
      contextParts.push(shortTermContext);
    }

    if (includeLongTerm && query) {
      const relevantMemories = await this.longTerm.search(query, 3);

      if (relevantMemories.length > 0) {
        const memoriesText = relevantMemories.map(mem => {
          return `- ${mem.content} (重要性：${mem.importance.toFixed(2)})`;
        }).join('\n');
        contextParts.push(`\n[相关长期记忆]\n${memoriesText}`);
      }
    }

    return contextParts.join('\n\n');
  }

  /**
   * 保存重要事件
   * @param eventDescription 事件描述
   * @param importance 重要性（默认0.8）
   * @returns 记忆ID
   */
  async saveImportantEvent(
    eventDescription: string,
    importance: number = 0.8
  ): Promise<string> {
    return await this.longTerm.addMemory(
      eventDescription,
      importance,
      { type: "event" }
    );
  }

  /**
   * 压缩会话记忆
   */
  async compactSessionMemory(): Promise<void> {
    const messages = this.shortTerm.getMessages();
    const compactedMessages = this.sessionMemory.compactSessionMemory(messages);

    // 清空短期记忆
    this.shortTerm.clear();

    // 添加压缩后的消息，确保它们是有效的BaseMessage对象
    for (const msg of compactedMessages) {
      if (msg && typeof msg === 'object' && 'type' in msg && 'content' in msg) {
        this.shortTerm.addMessage(msg);
      }
    }
  }

  /**
   * 获取系统状态
   * @returns 状态对象
   */
  async getState(): Promise<Record<string, any>> {
    return {
      agent_id: this.agentId,
      short_term: this.shortTerm.getState(),
      long_term_count: await this.longTerm.count(),
      importance_threshold: this.importanceThreshold
    };
  }

  /**
   * 清空所有记忆
   */
  async clear(): Promise<void> {
    this.shortTerm.clear();
    await this.longTerm.clear();
    this.sessionMemory.clear();
  }
}

/**
 * 创建记忆系统
 * @param agentId 智能体ID
 * @param windowSize 窗口大小（默认5）
 * @param kwargs 额外参数
 * @returns 双记忆系统实例
 */
export function createMemorySystem(
  agentId: string,
  windowSize: number = 5,
  kwargs: Record<string, any> = {}
): DualMemorySystem {
  return new DualMemorySystem(
    agentId,
    windowSize,
    kwargs.longTermStoragePath
  );
}

/**
 * 创建PostgreSQL记忆系统
 * @param agentId 智能体ID
 * @param windowSize 窗口大小（默认5）
 * @param collectionName 集合名称（默认"default"）
 * @returns 双记忆系统实例
 */
export async function createPgMemorySystem(
  agentId: string,
  windowSize: number = 5,
  collectionName: string = "default"
): Promise<DualMemorySystem> {
  const system = new DualMemorySystem(
    agentId,
    windowSize
  );
  
  // 替换长期记忆后端为 PostgreSQL
  const pgMemory = new PGLongTermMemory(agentId, collectionName);
  await pgMemory.initialize();
  system.setLongTermMemory(pgMemory);
  
  return system;
}

/**
 * 创建ChromaDB记忆系统
 * @param agentId 智能体ID
 * @param windowSize 窗口大小（默认5）
 * @param collectionName 集合名称（默认"long_term_memory"）
 * @returns 双记忆系统实例
 */
export function createChromaMemorySystem(
  agentId: string,
  windowSize: number = 5,
  collectionName: string = "long_term_memory"
): DualMemorySystem {
  const system = new DualMemorySystem(
    agentId,
    windowSize
  );
  
  // 替换长期记忆后端为 ChromaDB
  const chromaMemory = new ChromaLongTermMemory(agentId, collectionName);
  system.setLongTermMemory(chromaMemory);
  
  return system;
}
